### PR TITLE
[WIP] ServiceManager v2 + v3 compatibility

### PR DESCRIPTION
**Pull request by ezimuel**

### CODE REVIEW

**Comment by ezimuel** [comment]
@weierophinney please review this PR, thanks!

File: composer.json [hunk 1]
```diff
@@ -14,13 +14,13 @@
     },
     "require": {
         "php": ">=5.5",
-        "zendframework/zend-stdlib": "~2.5",
-        "zendframework/zend-servicemanager": "dev-develop as 2.7.0",
-        "zendframework/zend-eventmanager": "dev-develop as 2.7.0"
+        "zendframework/zend-stdlib": "~2.7",
+        "zendframework/zend-servicemanager": "^2.7.3 || ^3.0",
+        "zendframework/zend-eventmanager": "^2.6.2 || ^3.0"
```
Comment (weierophinney) [inline composer.json]: w00t! we can use tagged versions!

File: composer.json [hunk 2]
```diff
@@ -14,13 +14,13 @@
...
     },
     "require-dev": {
         "zendframework/zend-serializer": "dev-develop as 2.6.0",
-        "zendframework/zend-session": "~2.5",
+        "zendframework/zend-session": "dev-develop as 2.6.0",
```
Comment (weierophinney) [inline composer.json]: Note to future selves: we'll want to update this to a tagged version once we've tagged a version containing the SM/EM updates.

File: src/Service/StorageCacheAbstractServiceFactory.php [hunk 3]
```diff
@@ -45,6 +45,16 @@ public function canCreateServiceWithName(ContainerInterface $container, $request
...
     /**
+     * @param  ServiceLocatorInterface $serviceLocator
+     * @param  string $name
+     * @param  string $requestedName
+     * @return boolean
+     */
+    public function canCreateServiceWithName(ServiceLocatorInterface $serviceLocator, $name, $requestedName)
+    {
+        return $this->canCreate($serviceLocator, $requestedName);
+    }
```
Comment (weierophinney) [inline src/Service/StorageCacheAbstractServiceFactory.php]: Add an empty line following this new method. (CS)

**Comment by weierophinney** [comment]
@ezimuel Looks great!

Please add the `.travis.yml` changes as seen in the [relevant zend-barcode PR](zendframework/zend-barcode#16) so that we can validate that the code works properly with both v2 and v3 versions of the libraries going forward. After we get the results from that change, we'll know for certain if it's ready!

File: src/PatternPluginManager.php [hunk 4]
```diff
@@ -42,7 +43,14 @@ class PatternPluginManager extends AbstractPluginManager
...
      *
-     * @var array
+     * @var boolean
+     */
+    protected $shareByDefault = false;
+
+    /**
+     * Don't share by default
+     *
+     * @var boolean
```
Comment (marc-mabe) [inline src/PatternPluginManager.php]: @ezimuel can you please write in the doc-block which property belongs to which service manager version as for me this looks like a mistake.

File: src/Storage/AdapterPluginManager.php [hunk 5]
```diff
@@ -77,7 +77,47 @@ class AdapterPluginManager extends AbstractPluginManager
     protected $sharedByDefault = false;
 
     /**
+     * Don't share by default
+     *
+     * @var boolean
+     */
+    protected $shareByDefault = false;
+
+    /**
```
Comment (marc-mabe) [inline src/Storage/AdapterPluginManager.php]: same here

**Comment by marc-mabe** [comment]
@ezimuel Why did you remove all `PECL_INSTALL_* environment variables in `.travis.yml`?

**Comment by marc-mabe** [comment]
@ezimuel @weierophinney status ?

**Comment by weierophinney** [comment]
Merged with #68 / #69 (which were rebased).

**Pull request closed without merging**